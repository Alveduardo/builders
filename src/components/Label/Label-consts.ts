export const LABEL = {
  TEST_ID: {
    CONTAINER: 'LABEL_CONTAINER_TEST_ID',
  },
  ALLOWED_PROPERTIES: ['width'],
};
