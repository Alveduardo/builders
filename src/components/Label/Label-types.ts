import { TextProps, TextStyle } from 'react-native';
import { Color } from '../../utils/colors/colors-types';

export type LabelSize =
  | 'xxSmall'
  | 'xSmall'
  | 'small'
  | 'regular'
  | 'large'
  | 'xLarge'
  | 'xxLarge'
  | 'xxxLarge'
  | 'display';

export type SizeStyles = {
  [key in LabelSize]: {
    textStyle: TextStyle;
  };
};

interface DefaultStyles {
  textStyle: TextStyle;
}

export interface LabelStyles {
  defaultStyles: DefaultStyles;
  sizeStyles: SizeStyles;
}

type StylesConfig = DefaultStyles;

export interface LabelConfig {
  styles: StylesConfig;
}

export interface LabelProps extends TextProps {
  /**
   * Label color values.
   */
  color?: Color;

  /**
   * Label size values.
   */
  size?: LabelSize;
}
