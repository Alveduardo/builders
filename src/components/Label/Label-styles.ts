import { LabelStyles } from './Label-types';

export const labelStyles: LabelStyles = {
    defaultStyles: {
        textStyle: {

        },
    },
    sizeStyles: {
        xSmall: {
            textStyle: {
                fontSize: 10
            },
        },
        small: {
            textStyle: {
                fontSize: 14,
            }
        },
        regular: {
            textStyle: {
                fontSize: 18,
            }
        },
        large: {
            textStyle: {
                fontSize: 24,
            }
        },
        xLarge: {
            textStyle: {
                fontSize: 32,
            }
        },
        display: {
            textStyle: {
                fontSize: 84
            }
        }
    },
};
