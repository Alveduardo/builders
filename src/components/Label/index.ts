import Label from './Label'
import { LabelProps, LabelStyles } from './Label-types'

export default Label
export { LabelProps, LabelStyles }