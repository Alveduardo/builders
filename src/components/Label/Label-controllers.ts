import { labelStyles } from './Label-styles';
import { LabelConfig, LabelProps } from './Label-types';

export const useLabelConfig = ({ color, size }: Partial<LabelProps>): LabelConfig => {
    const {
        defaultStyles: { textStyle },
        sizeStyles: {
            [size!]: { textStyle: sizeStyle },
        },
    } = labelStyles;

    textStyle.color = color

    return {
        styles: {
            textStyle: { ...textStyle, ...sizeStyle },
        },
    };
};
