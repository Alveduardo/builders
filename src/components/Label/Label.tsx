import React from 'react';
import { Text } from 'react-native';
import { LABEL } from './Label-consts';
import { useLabelConfig } from './Label-controllers';
import { LabelProps } from './Label-types';

const Label = ({ children, color = '#FFF', size = 'small', style = undefined, ...rest }: LabelProps): JSX.Element => {
  const {
    styles: { textStyle },
  } = useLabelConfig({ color, size, style });

  return (
    <Text testID={LABEL.TEST_ID.CONTAINER} {...{ rest, style: [textStyle, style] }}>
      {children}
    </Text>
  );
};

export default Label;
