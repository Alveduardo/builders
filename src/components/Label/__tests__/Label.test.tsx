import React from 'react';
import { render } from '@testing-library/react-native';
import Label from '../Label';
import { LabelSize } from '../Label-types';
import { StyleSheet } from 'react-native';
import { LABEL } from '../Label-consts';

const CONTENT = 'txt example';

const describeEachSize = [
  {
    size: 'xxSmall',
    fontSize: 12,
  },
  {
    size: 'xSmall',
    fontSize: 14,
  },
  {
    size: 'small',
    fontSize: 16,
  },
  {
    size: 'regular',
    fontSize: 18,
  },
  {
    size: 'large',
    fontSize: 24,
  },
  {
    size: 'xLarge',
    fontSize: 32,
  },
  {
    size: 'xxLarge',
    fontSize: 48,
  },
  {
    size: 'xxxLarge',
    fontSize: 64,
  },
  {
    size: 'display',
    fontSize: 84,
  },
];

describe('Label ->', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(jest.fn());
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });

  describe('Snapshot ->', () => {
    it('Deve corresponder ao componente com as props default', () => {
      const tree = render(<Label>{CONTENT}</Label>).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('Deve corresponder ao componente com as props mais utilizadas', () => {
      const tree = render(
        <Label size='regular' color='#FF0000'>
          {CONTENT}
        </Label>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('props ->', () => {
    describe('color ->', () => {
      it('Deve renderizar o componente alterando a prop color', () => {
        const color = '#FF0000';

        const { getByTestId } = render(<Label {...{ color }}>{CONTENT}</Label>);

        const containerStyle = StyleSheet.flatten(getByTestId(LABEL.TEST_ID.CONTAINER).props.style);

        expect(containerStyle.color).toBe(color);
      });
    });
    describe('size ->', () => {
      describe.each(describeEachSize)('Para cada valor de size ->', ({ size, fontSize }) => {
        it(`Deve ter o valor de ${fontSize} de acordo com a prop ${size}`, () => {
          const { getByTestId } = render(<Label size={size as LabelSize}>{CONTENT}</Label>);

          const containerStyle = StyleSheet.flatten(getByTestId(LABEL.TEST_ID.CONTAINER).props.style);
          expect(containerStyle.fontSize).toBe(fontSize);
        });
      });
    });
    describe('children ->', () => {
      it('Deve renderizar o componente exibindo o children', () => {
        const color = '#FF0000';

        const { getByTestId } = render(<Label {...{ color }}>{CONTENT}</Label>);

        const componente = getByTestId(LABEL.TEST_ID.CONTAINER).props;

        expect(componente.children).toEqual(CONTENT);
      });
    });
  });
});
