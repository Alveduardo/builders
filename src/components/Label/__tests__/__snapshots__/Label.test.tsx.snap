// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label -> Snapshot -> Deve corresponder ao componente com as props default 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#FFF",
        "fontSize": 16,
      },
      undefined,
    ]
  }
  testID="LABEL_CONTAINER_TEST_ID"
>
  txt example
</Text>
`;

exports[`Label -> Snapshot -> Deve corresponder ao componente com as props mais utilizadas 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#FF0000",
        "fontSize": 18,
      },
      undefined,
    ]
  }
  testID="LABEL_CONTAINER_TEST_ID"
>
  txt example
</Text>
`;
